lmnet/lmnet/networks/keypoint_detection/base.py

```diff
@@ -0,0 +1,205 @@
+from lmnet.post_processor import gaussian_heatmap_to_joints
+from lmnet.visualize import visualize_keypoint_detection
+
+
+class Base(BaseNetwork):
+    """base network for keypoint detection
+
+    This base network is for keypoint detection.
+    Each keypoint detection network class should extend this class.
+
+    """
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+
+    def placeholders(self):
+
+        if self.data_format == 'NHWC':
+            shape = (self.batch_size, self.image_size[0], self.image_size[1], 3)
+        else:
+            shape = (self.batch_size, 3, self.image_size[0], self.image_size[1])
+
+        images_placeholder = tf.placeholder(tf.float32, shape=shape, name="images_placeholder")
+        labels_placeholder = tf.placeholder(
+            tf.float32,
+            shape=(self.batch_size,
+                   self.image_size[0] // self.stride, self.image_size[1] // self.stride,
+                   self.num_joints),
+            name="labels_placeholder")
+
+        return images_placeholder, labels_placeholder
+
+    def inference(self, images, is_training):
+        base = self.base(images, is_training)
+        return tf.identity(base, name="output")
+
+    def _colored_heatmaps(self, heatmaps, color, name=""):
+        """Visualize heatmaps with given color.
+
+        Args:
+            heatmaps: a Tensor of shape (batch_size, height, width, num_joints).
+            color: a numpy array of shape (batch_size, 1, 1, num_joints, 3).
+            name: str, name to display on tensorboard.
+
+        """
+
+        heatmaps_colored = tf.expand_dims(heatmaps, axis=-1)
+        heatmaps_colored *= color
+        heatmaps_colored = tf.reduce_sum(heatmaps_colored, axis=3)
+
+        tf.summary.image(name, heatmaps_colored)
+
+    @staticmethod
+    def py_post_process(heatmaps, num_dimensions=2, stride=2):
+        """Convert from heatmaps to joints, it is mainly used for visualization and metrics in training time.
+
+        Args:
+            heatmaps: a numpy array of shape (batch_size, height, width, num_joints).
+            num_dimensions: int.
+            stride: int, stride = image_height / heatmap_height.
+
+        Returns:
+            batch_joints: a numpy array of shape (batch_size, num_joints, 3).
+
+        """
+        batch_size = heatmaps.shape[0]
+        list_joints = [gaussian_heatmap_to_joints(heatmaps[i], num_dimensions, stride=stride)
+                       for i in range(batch_size)]
+
+        return np.stack(list_joints)
+
+    def post_process(self, output):
+        """Tensorflow mirror method for py_post_process(),
+        it is mainly used for visualization and metrics in training time.
+
+        Args:
+            output: a Tensor of shape (batch_size, height, width, num_joints).
+
+        Returns:
+            joints: a Tensor of shape (batch_size, num_joints, 3).
+
+        """
+
```

Comment:
Could you remove this blank line? There may or may not be a blank line immediately after the docstring for the method. I want to eliminate it gradually to condense the code vertically.

Also, could you make the same modification for all method?